# Segmenting and Clustering Neighborhoods in Toronto

### This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers

### Installing and importing requested libraries

In [1]:
import pandas as pd
import requests

## Question 1

In [2]:
wiki_data = pd.read_html(requests.get('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969').text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [3]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

### Data preprocessing and clening

In [4]:
# Dropping the rows where Borough is 'Not assigned'
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [5]:
### Grouping the records based on postal code
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [6]:
# Checking for number of records where Neighbourhood is "Not assigned"
df.Neighbourhood.str.count("Not assigned").sum()

0

In [7]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [8]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.shape

(103, 3)

 Answer to Question 1: We have 103 rows and 3 columns

## Question 2

installing geocoder

In [10]:
pip install geocoder

     |████████████████████████████████| 98 kB 9.2 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [11]:
import geocoder

In [12]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [13]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [14]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [15]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [16]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [17]:
combined_data.shape

(103, 5)

Answer to Question 2: We get 103 rows as expected when we do a inner join, so we have good data.

## Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [18]:
from geopy.geocoders import Nominatim

In [19]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Lets visualize the map of Toronto

In [20]:
import folium

In [21]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [22]:
CLIENT_ID = 'FXW45F0JDJVAXMSJSJWDHICQU5AG0WAAL4KPGOJH13N3KJT3' 
CLIENT_SECRET = 'N0DMVTAEFPSD4LEVR4SLUWYLIYFU21SFVTPL4JR5GNP4GMIV'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FXW45F0JDJVAXMSJSJWDHICQU5AG0WAAL4KPGOJH13N3KJT3
CLIENT_SECRET:N0DMVTAEFPSD4LEVR4SLUWYLIYFU21SFVTPL4JR5GNP4GMIV


Next, we create a function to get all the venue categories in Toronto

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [24]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [25]:
venues_in_toronto.shape

(1265, 5)

In [26]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                    Venue        Venue Category  
0         Brookbanks Park                  Park  
1                     KFC  Fast Food Restaurant  
2             649 Variety     Convenience Store  
3           Variety Store     Food & Drink Shop  
4  Victoria Village Arena          Hockey Arena

In [27]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                             Parkwoods   
4                                      Victoria Village   
...                                                 ...   
1253  Mimico NW, The Queensway West, South of Bloor,...   
1254  Mimico NW, The Queensway West, South of Bloor,...   
1255  Mimico NW, The Queensway West, South of Bloor,...   
1256  Mimico NW, The Queensway West, South of Bloor,...   
1257  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656                     KFC   
2                  43.753259               -79.329656             649 Variety   
3                  43.753259               -79.329656           Variety Store   
4                  43.725882               -79.315572  Victoria Village Arena   
...                      ...                      ...                     ...   
1253               43.628841               -79.520999              Wingporium   
1254               43.628841               -79.520999        South St. Burger   
1255               43.628841               -79.520999               Dollarama   
1256               43.628841               -79.520999          Healthy Planet   
1257               43.628841               -79.520999                7-Eleven   

            Venue Category  
0                     Park  
1     Fast Food Restaurant  
2        Convenience Store  
3        Food & Drink Shop  
4             Hockey Arena  
...                    ...  
1253           Wings Joint  
1254          Burger Joint  
1255        Discount Store  
1256       Supplement Shop  
1257     Convenience Store  

[417 rows x 5 columns]

Checking for the maximum venue categories

In [28]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                              The Danforth West, Riverdale   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.679557               -79.340923   

                                              Venue  
Venue Category                                       
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[225 rows x 4 columns]

# One Hot encoding the venue Categories

In [29]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   
...              ...      ...                 ...           ...   
1260               0        0                   0             0   
1261               0        0                   0             0   
1262               0        0                   0             0   
1263               0        0                   0             0   
1264               0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1260               0                0                 0                    0   
1261               0                0                 0                    0   
1262               0                0                 0                    0   
1263               0                0                 0                    0   
1264               0                0                 0                    0   

      Aquarium  Art Gallery  ...  Train Station  Turkish Restaurant  \
0            0            0  ...              0                   0   
1            0            0  ...              0                   0   
2            0            0  ...              0                   0   
3            0            0  ...              0                   0   
4            0            0  ...              0                   0   
...        ...          ...  ...            ...                 ...   
1260         0            0  ...              0                   0   
1261         0            0  ...              0                   0   
1262         0            0  ...              0                   0   
1263         0            0  ...              0                   0   
1264         0            0  ...              0                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1260                              0                 0                      0   
1261                              0                 0                      0   
1262                              0                 0                      0   
1263                              0                 0                      0   
1264                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0       

Adding the neighbourhood to the encoded dataframe

In [30]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Adult Boutique  Airport  Airport Food Court  \
0         Parkwoods               0        0                   0   
1         Parkwoods               0        0                   0   
2         Parkwoods               0        0                   0   
3         Parkwoods               0        0                   0   
4  Victoria Village               0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Aquarium  ...  Train Station  Turkish Restaurant  \
0                    0         0  ...              0                   0   
1                    0         0  ...              0                   0   
2                    0         0  ...              0                   0   
3                    0         0  ...              0                   0   
4                    0         0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 226 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [31]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Adult Boutique  Airport  \
0                                        Agincourt             0.0      0.0   
1                           Alderwood, Long Branch             0.0      0.0   
2  Bathurst Manor, Wilson Heights, Downsview North             0.0      0.0   
3                                  Bayview Village             0.0      0.0   
4                Bedford Park, Lawrence Manor East             0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Aquarium  ...  Train Station  \
0               0.0             0.000000       0.0  ...            0.0   
1               0.0             0.000000       0.0  ...            0.0   
2               0.0             0.000000       0.0  ...            0.0   
3               0.0             0.000000       0.0  ...            0.0   
4               0.0             0.043478       0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 226 columns]

Let's make a function to get the top most common venue categories

In [32]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [33]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East            Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant        Clothing Store   
1                  Pool                        Gym              Pharmacy   
2           Coffee Shop                       Park        Ice Cream Shop   
3                  Bank        Japanese Restaurant                  Café   
4           Coffee Shop             Sandwich Place  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Skating Rink        Adult Boutique                 Motel   
1          Skating Rink           Coffee Shop        Sandwich Place   
2         Shopping Mall        Sandwich Place                 Diner   
3        Adult Boutique         Movie Theater        Medical Center   
4      Sushi Restaurant                   Pub                  Café   

      8th Most Common Venue      9th Most Common Venue  \
0  Mediterranean Restaurant                Men's Store   
1                       Pub  Middle Eastern Restaurant   
2          Sushi Restaurant              Deli / Bodega   
3  Mediterranean Restaurant                Men's Store   
4        Italian Restaurant               Liquor Store   

       10th Most Common Venue  
0               Metro Station  
1  Modern European Restaurant  
2                 Bridal Shop  
3               Metro Station  
4             Thai Restaurant

Let's make the model to cluster our Neighbourhoods

In [35]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [36]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [37]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0,
       3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 3, 0, 0, 0, 3,
       0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       3, 0, 3, 0, 0, 0, 0, 3, 1], dtype=int32)

In [38]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [39]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             3.0                  Park   
1  43.725882 -79.315572             0.0           Pizza Place   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
0  Fast Food Restaurant       Convenience Store      Food & Drink Shop   
1          Intersection             Coffee Shop  Portuguese Restaurant   
2                Bakery                    Park                   Café   
3           Coffee Shop  Furniture / Home Store     Miscellaneous Shop   
4      Sushi Restaurant           Burrito Place            Yoga Studio   

   5th Most Common Venue            6th Most Common Venue  \
0         Adult Boutique  Molecular Gastronomy Restaurant   
1           Hockey Arena                   Medical Center   
2                    Pub                      Event Space   
3  Vietnamese Restaurant                         Boutique   
4         Sandwich Place                    Smoothie Shop   

      7th Most Common Venue 8th Most Common Venue     9th Most Common Venue  \
0            Massage Studio        Medical Center  Mediterranean Restaurant   
1  Mediterranean Restaurant           Men's Store             Metro Station   
2            Breakfast Spot    Mexican Restaurant       Distribution Center   
3               Men's Store         Metro Station                    Museum   
4              Burger Joint    Mexican Restaurant                      Café   

  10th Most Common Venue  
0            Men's Store  
1     Mexican Restaurant  
2      French Restaurant  
3     Mexican Restaurant  
4            Salad Place

In [40]:
# Drop all NaN values
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [41]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [42]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [43]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             0.0           Pizza Place   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
6         Scarborough             0.0  Fast Food Restaurant   
..                ...             ...                   ...   
97   Downtown Toronto             0.0                  Café   
98          Etobicoke             0.0                 River   
99   Downtown Toronto             0.0      Sushi Restaurant   
100      East Toronto             0.0    Light Rail Station   
102         Etobicoke             0.0         Tanning Salon   

    2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
1            Intersection             Coffee Shop  Portuguese Restaurant   
2                  Bakery                    Park                   Café   
3             Coffee Shop  Furniture / Home Store     Miscellaneous Shop   
4        Sushi Restaurant           Burrito Place            Yoga Studio   
6              Print Shop          Adult Boutique                  Motel   
..                    ...                     ...                    ...   
97             Restaurant                     Gym            Coffee Shop   
98             Smoke Shop                    Pool                  Motel   
99    Japanese Restaurant          Adult Boutique              Bookstore   
100  Gym / Fitness Center           Garden Center             Skate Park   
102       Supplement Shop           Grocery Store            Wings Joint   

     5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
1             Hockey Arena        Medical Center  Mediterranean Restaurant   
2                      Pub           Event Space            Breakfast Spot   
3    Vietnamese Restaurant              Boutique               Men's Store   
4           Sandwich Place         Smoothie Shop              Burger Joint   
6       Mac & Cheese Joint        Massage Studio            Medical Center   
..                     ...                   ...                       ...   
97               Speakeasy                 Hotel            Sandwich Place   
98          Massage Studio        Medical Center  Mediterranean Restaurant   
99             Men's Store       Bubble Tea Shop             Burrito Place   
100             Comic Shop        Farmers Market      Fast Food Restaurant   
102                 Bakery        Hardware Store         Convenience Store   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                 Men's Store         Metro Station     Mexican Restaurant  
2          Mexican Restaurant   Distribution Center      French Restaurant  
3               Metro Station                Museum     Mexican Restaurant  
4          Mexican Restaurant                  Café            Salad Place  
6    Mediterranean Restaurant           Men's Store          Metro Station  
..                        ...                   ...                    ...  
97                 Steakhouse                   Pub               Tea Room  
98                Men's Store         Metro Station     Mexican Restaurant  
99         Mexican Restaurant  Ethiopian Restaurant            Escape Room  
100             Burrito Place            Restaurant             Smoke Shop  
102              Burger Joint         Burrito Place         Discount Store  

[89 rows x 12 columns]

cluster 2

In [44]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             1.0             Cafeteria                Lounge   

   3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
45        Massage Studio        Medical Center  Mediterranean Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45           Men's Store         Metro Station    Mexican Restaurant   

        9th Most Common Venue 10th Most Common Venue  
45  Middle Eastern Restaurant     Miscellaneous Shop

cluster 3

In [45]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field        Adult Boutique   
101   Etobicoke             2.0        Baseball Field        Adult Boutique   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57                  Motel    Mac & Cheese Joint        Massage Studio   
101                 Motel    Mac & Cheese Joint        Massage Studio   

    6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
57         Medical Center  Mediterranean Restaurant           Men's Store   
101        Medical Center  Mediterranean Restaurant           Men's Store   

    9th Most Common Venue 10th Most Common Venue  
57          Metro Station     Mexican Restaurant  
101         Metro Station     Mexican Restaurant

cluster 4

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             3.0                  Park   
21              York             3.0                  Park   
35         East York             3.0                  Park   
52        North York             3.0                  Park   
64              York             3.0                  Park   
66        North York             3.0                  Park   
77         Etobicoke             3.0                  Park   
91  Downtown Toronto             3.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue  \
0   Fast Food Restaurant     Convenience Store   
21         Women's Store                  Pool   
35     Convenience Store          Intersection   
52        Adult Boutique                 Motel   
64     Convenience Store        Adult Boutique   
66     Convenience Store        Adult Boutique   
77        Sandwich Place        Adult Boutique   
91            Playground                 Trail   

              4th Most Common Venue 5th Most Common Venue  \
0                 Food & Drink Shop        Adult Boutique   
21                   Adult Boutique                 Motel   
35                   Adult Boutique                 Motel   
52               Mac & Cheese Joint        Massage Studio   
64                            Motel    Mac & Cheese Joint   
66                            Motel    Mac & Cheese Joint   
77  Molecular Gastronomy Restaurant    Mac & Cheese Joint   
91                   Adult Boutique                 Motel   

              6th Most Common Venue     7th Most Common Venue  \
0   Molecular Gastronomy Restaurant            Massage Studio   
21               Mac & Cheese Joint            Massage Studio   
35                   Massage Studio            Medical Center   
52                   Medical Center  Mediterranean Restaurant   
64                   Massage Studio            Medical Center   
66                   Massage Studio            Medical Center   
77                   Massage Studio            Medical Center   
91               Mac & Cheese Joint            Massage Studio   

       8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
0             Medical Center  Mediterranean Restaurant            Men's Store  
21            Medical Center  Mediterranean Restaurant            Men's Store  
35  Mediterranean Restaurant               Men's Store          Metro Station  
52               Men's Store             Metro Station     Mexican Restaurant  
64  Mediterranean Restaurant               Men's Store          Metro Station  
66  Mediterranean Restaurant               Men's Store          Metro Station  
77  Mediterranean Restaurant               Men's Store          Metro Station  
91            Medical Center  Mediterranean Restaurant            Men's Store

We have successfully cluster Toronto neighbourhood based on venue categories!